internal/handlers/staking_restrictions_hooks.go

```diff
@@ -0,0 +1,141 @@
+	DefaultMaxCapPercent = 0.33
+	// Minimum Allowed Cap for Bonded stake of any single validator
+	DefaultMinCapPercent = 0.05
+)
+
+type RestrictionOptions struct {
+	MaxConcentrationMultiple float32
+	MaxBondedCapPercent      float32
+	MinBondedCapPercent      float32
+}
+
+// DefaultRestictionOptions are default constraints that prevent single point of failure on validators
+var DefaultRestictionOptions = &RestrictionOptions{
+	MaxConcentrationMultiple: DefaultConcentration,
+	MaxBondedCapPercent:      DefaultMaxCapPercent,
+	MinBondedCapPercent:      DefaultMinCapPercent,
+}
+
+// UnlimitedRestrictionOptions are used to remove restrictions for validator staking limits from delegations
+var UnlimitedRestrictionOptions = &RestrictionOptions{
+	MaxConcentrationMultiple: 1.0,
+	MaxBondedCapPercent:      1.0,
+	MinBondedCapPercent:      1.0,
+}
+
+// Hooks wrapper struct for slashing keeper
+type StakingRestrictionHooks struct {
+	k    *stakingkeeper.Keeper
+	opts RestrictionOptions
+}
+
+// NewStakingRestrictionHooks configures a hook that validates changes to delegation modifications and
+// prevents concentration of voting power beyond configured limits on active validators.
+func NewStakingRestrictionHooks(k *stakingkeeper.Keeper, opts RestrictionOptions) StakingRestrictionHooks {
+	return StakingRestrictionHooks{k, opts}
+}
+
+// Verifies that the delegation would not cause the validator's voting power to exceed our staking distribution limits
+func (h StakingRestrictionHooks) AfterDelegationModified(ctx sdktypes.Context, delAddr sdktypes.AccAddress, valAddr sdktypes.ValAddress) error {
+	valCount := len(h.k.GetLastValidators(ctx))
+
+	// do not bother with limits on networks this small (or under simulation).
+	if valCount < 4 || ctx.ChainID() == helpers.SimAppChainID {
+		return nil
+	}
+
+	// bond limit is allowed to have a multiple of even shares of network bonded stake.
+	maxValidatorPercent := h.opts.MaxConcentrationMultiple * (1.0 / float32(valCount))
+
+	// check the capped bond amount is within the overall range limits.
+	if maxValidatorPercent > h.opts.MaxBondedCapPercent {
+		maxValidatorPercent = h.opts.MaxBondedCapPercent
+	} else if maxValidatorPercent < h.opts.MinBondedCapPercent {
+		maxValidatorPercent = h.opts.MinBondedCapPercent
+	}
+
+	oldPower := h.k.GetLastValidatorPower(ctx, valAddr)
+	validator, found := h.k.GetValidator(ctx, valAddr)
+	if found {
+		power := validator.GetConsensusPower(h.k.PowerReduction(ctx))
+		currentBond := h.k.TotalBondedTokens(ctx)
+		maxBond := currentBond.Quo(math.NewInt(100)).MulRaw(int64(maxValidatorPercent * 100))
+
+		// if the power of this validator is increasing and it is over the maximum bonded token amount then we error out this transaction.
+		if power > oldPower && validator.GetBondedTokens().GT(maxBond) {
+			return sdkerrors.Wrapf(
+				sdkerrors.ErrInvalidRequest,
+				"validator bonded tokens of %d exceeds max of %d (%.1f%%) for %d validators",
```

Comment:
Probably want 2 digits after that decimal since we're using *100 for the `maxBond` calc.
```suggestion
				"validator bonded tokens of %d exceeds max of %d (%.2f%%) for %d validators",
```

Reply:
Oh... never mind.  Those two digits would be before the decimal.